torch/lib/c10d/TCPStore.hpp

```diff
@@ -36,29 +61,49 @@ class TCPStoreDaemon {
   void getNumKeysHandler(int socket) const;
   void deleteHandler(int socket);
   void waitHandler(int socket);
+  void watchHandler(int socket);
 
   bool checkKeys(const std::vector<std::string>& keys) const;
   void wakeupWaitingClients(const std::string& key);
+  void sendKeyUpdatesToClients(
+      const std::string& key,
+      const enum WatchResponseType& type,
+      std::vector<uint8_t>& oldData,
+      std::vector<uint8_t>& newData);
 
-  void initStopSignal();
-  void closeStopSignal();
-
-  std::thread daemonThread_;
+ private:
   std::unordered_map<std::string, std::vector<uint8_t>> tcpStore_;
-  // From key -> the list of sockets waiting on it
+  // From key -> the list of sockets waiting on the key
   std::unordered_map<std::string, std::vector<int>> waitingSockets_;
   // From socket -> number of keys awaited
   std::unordered_map<int, size_t> keysAwaited_;
+  // From key -> the list of sockets watching the key
+  std::unordered_map<std::string, std::vector<int>> watchedSockets_;
+};
 
-  std::vector<int> sockets_;
-  int storeListenSocket_;
-#ifdef _WIN32
-  const std::chrono::milliseconds checkTimeout_
-      = std::chrono::milliseconds(10);
-  HANDLE ghStopEvent_;
-#else
-  std::vector<int> controlPipeFd_{-1, -1};
-#endif
+// Listener thread runs on all processes
+// Right now only handles callbacks registered from watchKey()
+class ListenThread : public BackgroundThread {
```

Comment:
Is this specifically added for watch callbacks? If so, does it make sense to use the following class names?

BackgroundThread -> Daemon
ListenThread -> WatchCallbackDaemon

Reply:
Sorry, I missed this:

> Right now only handles callbacks registered from watchKey()

If we plan to add more features into this in the future, does it make sense to name them 

TCPStoreDaemon -> TCPStoreMasterDaemon
ListenThread -> TCPStoreWorkerDaemon

Reply:
Good idea, I will rename it to these.